[RISCV] Fix alignment when mixing rvc/norvc relax/norelax code

kito-cheng opened the pull request:
The issue we want to show is the alignment problem when mixing RVC and non-RVC code, especially when the RVC code is relaxed and the alignment with in the non-RVC region will not got expected alignment requirement.

Give an example, this example will build with `rv64gc` and `-mrelax` option, and assume text section is start from `0x1000`:

```
_start:
        lui a0, %hi(foo)
        addi a0, a0, %lo(foo)
        mul a0, a1, a4
        .option push

        .option norelax
        .option norvc
        .balign 8
SHOULD_ALIGN_8_HERE:
        .word 0x12345678
```

Then when we assemble this file, assembler will insert NOP and R_RISCV_RELAX/R_RISCV_ALIGN like below:
```
Disassembly of section .text:

0000000000000000 <_start-0x2>:
   0:   0001                    nop
                        0: R_RISCV_ALIGN        *ABS*+0x2

0000000000000002 <_start>:
   2:   00000537                lui     a0,0x0
                        2: R_RISCV_HI20 foo
                        2: R_RISCV_RELAX        *ABS*
   6:   00050513                mv      a0,a0
                        6: R_RISCV_LO12_I       foo
                        6: R_RISCV_RELAX        *ABS*
   a:   8082                    ret
   c:   00000013                nop

0000000000000010 <SHOULD_ALIGN_8_HERE>:
  10:   12345678                .word   0x12345678
```

And we didn't insert R_RISCV_ALIGN before `SHOULD_ALIGN_8_HERE`, because norelax option are given and the alignment seems already meet.

However...the linker relaxation will remove first NOP in the text section before the _start symbol for meet the alignment requirement, then `SHOULD_ALIGN_8_HERE` no longer aligned to 8 bytes.

So this should be fixed in the MC assembler by inserting R_RISCV_ALIGN correctly before `SHOULD_ALIGN_8_HERE`, even if the norelax option is given, but that not means we should always inserting that, we just need insert when relax has ever enabled on that section.

Also it should emit N-2 bytes NOPs even norvc status if this section has enable RVC ever, so that linker has enough NOPs to remove for meet the alignment requirement.

This is not really lld bug, but this must be put within lld testcases so that we can verify the alignment result is wrong after linking.

Not all testcase are fail with out this fix, however some of them may fail on binutils side, so I think it worth to put all testcase to lld test to prevent future regression.

kito-cheng committed:
[RISCV] Fix alignment when mixing rvc/norvc relax/norelax code

The issue we want to show is the alignment problem when mixing RVC and non-RVC
code, especially when the RVC code is relaxed and the alignment with in the
non-RVC region will not got expected alignment requirement.

Give an example, this example will build with `rv64gc` and `-mrelax` option,
and assume text section is start from `0x1000`:

```
_start:
        lui a0, %hi(foo)
        addi a0, a0, %lo(foo)
        mul a0, a1, a4
        .option push

        .option norelax
        .option norvc
        .balign 8
SHOULD_ALIGN_8_HERE:
        .word 0x12345678
```

Then when we assemble this file, assembler will insert NOP and
`R_RISCV_RELAX`/`R_RISCV_ALIGN` like below:
```
Disassembly of section .text:

0000000000000000 <_start-0x2>:
   0:   0001                    nop
                        0: R_RISCV_ALIGN        *ABS*+0x2

0000000000000002 <_start>:
   2:   00000537                lui     a0,0x0
                        2: R_RISCV_HI20 foo
                        2: R_RISCV_RELAX        *ABS*
   6:   00050513                mv      a0,a0
                        6: R_RISCV_LO12_I       foo
                        6: R_RISCV_RELAX        *ABS*
   a:   8082                    ret
   c:   00000013                nop

0000000000000010 <SHOULD_ALIGN_8_HERE>:
  10:   12345678                .word   0x12345678
```

And we didn't insert `R_RISCV_ALIGN` before `SHOULD_ALIGN_8_HERE`, because
`norelax` option are given and the alignment seems already meet.

However...the linker relaxation will remove first NOP in the text section
before the `_start` symbol for meet the alignment requirement, then
`SHOULD_ALIGN_8_HERE` no longer aligned to 8 bytes.

So this should be fixed in the MC assembler by inserting `R_RISCV_ALIGN`
correctly before `SHOULD_ALIGN_8_HERE`, even if the norelax option is given,
but that not means we should always inserting that, we just need insert when
relax has ever enabled on that section.

Also it should emit N-2 bytes NOPs even norvc status if this section has enable
RVC ever, so that linker has enough NOPs to remove for meet the alignment
requirement.

This is not really lld bug, but this must be put within lld testcases so that we
can verify the alignment result is wrong after linking.

Not all testcase are fail with out this fix, however some of them may fail on
binutils side, so I think it worth to put all testcase to lld test to prevent
future regression.

topperc commented on llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:
```diff
@@ -3177,6 +3177,18 @@ bool RISCVAsmParser::parseDirectiveOption() {
+    auto *Section = Streamer.getCurrentSectionOnly();
+    // Update RVCEver status only if any instruction emitted.
+    if (Section->hasInstructions() && Assembler != nullptr &&
+        (getSTI().hasFeature(RISCV::FeatureStdExtC) ||
```
C implies Zca

topperc commented:
Does the same issue exist with .option exact?

MaskRay commented:
I think the example is missing a `.balign 4` at `_start`. Without the alignment, the initial alignment is 2, <= the minimum nop size (2 due to the C extension), and assemblers (both GAS and LLVM) will not emit a R_RISCV_ALIGN relocation.

ELF assembles have the subsection feature. While I haven't tested, I think the patch will not work with

```
        .text 1
        .option push
        .option norelax
        .option norvc
        .balign 8
SHOULD_ALIGN_8_HERE:
        .word 0x12345678
        .option pop

        .text 0
_start:
.balign 4
        lui a0, %hi(foo)
        addi a0, a0, %lo(foo)
        mul a0, a1, a4
```

I think we need to modify MCSection::LinkerRelaxable. It should be set to a value that indicates the `LayoutOrder` of the first linker-relaxable fragment. Then, relaxAlign should be updated to verify if the FT_Align fragment appears after this first linker-relaxable fragment.

MaskRay commented:
> I think we need to modify MCSection::LinkerRelaxable. It should be set to a value that indicates the `LayoutOrder` of the first linker-relaxable fragment. Then, relaxAlign should be updated to verify if the FT_Align fragment appears after this first linker-relaxable fragment.

See #150816

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -309,9 +309,12 @@ bool RISCVAsmBackend::relaxAlign(MCFragment &F, unsigned &Size) {
   // is larger than the nop size.
   const MCSubtargetInfo *STI = F.getSubtargetInfo();
-  if (!STI->hasFeature(RISCV::FeatureRelax))
+  if (!hasRelaxEver(F.getParent()) && !STI->hasFeature(RISCV::FeatureRelax))
```
I assume `STI->hasFeature(RISCV::FeatureRelax)` is usually true so we should check it first before doing the set lookup

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.h:
```diff
@@ -33,11 +35,23 @@ class RISCVAsmBackend : public MCAsmBackend {
   ~RISCVAsmBackend() override = default;
 
+  void setRVCEver(MCSection *Sec) { RVCEverSections.insert(Sec); }
+  bool hasRVCEver(MCSection *Sec) const {
```
hasRVCEver is never called.

kito-cheng committed:
!fixup

Changes:
- Add const to MCSection* for the RelaxEverSections and RVCEverSections
- Fix and simplify condition in RISCVAsmBackend::relaxAlign.
- Drop lld tests.
- Adding MC tests

kito-cheng commented:
Changes:
- Add `const` to `MCSection*` for the `RelaxEverSections` and `RVCEverSections`
- Fix and simplify condition in `RISCVAsmBackend::relaxAlign`.
- Drop lld tests.
- Adding MC tests

lenary commented:
I don't know of the status of this vs #150816. I presume we're only getting one of these, not both.

In terms of naming, I would prefer we don't reference rvc, and instead we reference this in terms of the minimal `IALIGN` - which for any section will be 4 until someone enables Zca at which point it will drop to 2. I think conceptually that's a bit more specific than noting something about "rvc" when the check is for zca, a subset of rvc. I'm happy to accept the code would be equivalent, but I think naming in terms of `IALIGN` would help clarify what's really going on.

kito-cheng commented:
I believe https://github.com/llvm/llvm-project/pull/150816 is better solution than mine, and I am happy to see that moving forward, so let me close this PR by myself to prevent confusing people that we are competing :)
